GPU/TPCFastTransformation/Spline1DSpec.h

```diff
@@ -0,0 +1,528 @@
+  void moveBufferTo(char* newBufferPtr);
+#endif
+
+  using FlatObject::releaseInternalBuffer;
+
+  void destroy();
+  void setActualBufferAddress(char* actualFlatBufferPtr);
+  void setFutureBufferAddress(char* futureFlatBufferPtr);
+
+ protected:
+  /// Non-const accessor to the knots array
+  Knot* getKnots() { return reinterpret_cast<Knot*>(mFlatBufferPtr); }
+
+  /// Non-const accessor to U->knots map
+  int* getUtoKnotMap() { return mUtoKnotMap; }
+
+#if !defined(GPUCA_GPUCODE)
+  /// Constructor for a regular spline
+  void recreate(int nYdim, int numberOfKnots);
+
+  /// Constructor for an irregular spline
+  void recreate(int nYdim, int numberOfKnots, const int knotU[]);
+#endif
+
+  /// _____________  Data members  ____________
+
+  int mYdim = 0;                ///< dimentionality of F
+  int mNumberOfKnots = 0;       ///< n knots on the grid
+  int mUmax = 0;                ///< U of the last knot
+  DataT mXmin = 0;              ///< X of the first knot
+  DataT mXtoUscale = 0;         ///< a scaling factor to convert X to U
+  int* mUtoKnotMap = nullptr;   //! (transient!!) pointer to (integer U -> knot index) map inside the mFlatBufferPtr array
+  DataT* mParameters = nullptr; //! (transient!!) pointer to F-dependent parameters inside the mFlatBufferPtr array
+
+#ifndef GPUCA_ALIROOT_LIB
+  ClassDefNV(Spline1DContainer, 1);
+#endif
+};
+
+template <typename DataT>
+template <typename Spline1DContainer<DataT>::SafetyLevel SafeT>
+GPUdi() int Spline1DContainer<DataT>::getLeftKnotIndexForU(DataT u) const
+{
+  /// Get i: u is in [knot_i, knot_{i+1}) segment
+  /// when u is otside of [0, mUmax], return a corresponding edge segment
+  int iu = (int)u;
+  if (SafeT == SafetyLevel::kSafe) {
+    iu = (iu < 0) ? 0 : (iu > mUmax ? mUmax : iu);
+  }
+  return getUtoKnotMap()[iu];
+}
+
+template <typename DataT>
+GPUdi() void Spline1DContainer<DataT>::setXrange(DataT xMin, DataT xMax)
+{
+  mXmin = xMin;
+  double l = ((double)xMax) - xMin;
+  if (l < 1.e-8) {
+    l = 1.e-8;
+  }
+  mXtoUscale = mUmax / l;
+}
+
+/// ==================================================================================================
+/// The specification declares common methods for all other Spline1D specifications.
+/// Implementations of the methods may depend on the YdimT value.
+///
+template <typename DataT, int YdimT, bool YisPositiveT, bool YisOneT, bool YisAbsentT>
+class Spline1DSpec<DataT, YdimT, true /*YisAnyT*/, YisPositiveT, YisOneT, YisAbsentT> : public Spline1DContainer<DataT>
+{
+  typedef Spline1DContainer<DataT> TBase;
+
+ public:
+  typedef typename TBase::SafetyLevel SafetyLevel;
+  typedef typename TBase::Knot Knot;
+
+  /// _______________  Template magic  ________________________
+
+  /// An expression getYdim(YisPositiveTbool{},int) is either an integer or a constexpr integer,
+  /// depending on the YisPositiveT value
+  ///
+  typedef std::integral_constant<bool, YisPositiveT> YisPositiveTbool;
```

Comment:
GPU compilation on OpenCL fails due to usage of the std namespace.
All system headers and usage of `std::` should be masked with `#ifndef GPUCA_GPUCODE_DEVICE`:
```
/home/qon/alice/O2/GPU/GPUTracking/Standalone/TPCFastTransformation/Spline1DSpec.h:284:11: error: use of undeclared identifier 'std'
  typedef std::integral_constant<bool, YisPositiveT> YisPositiveTbool;
          ^
/home/qon/alice/O2/GPU/GPUTracking/Standalone/TPCFastTransformation/Spline1DSpec.h:285:32: error: use of undeclared identifier 'std'
  static constexpr int getYdim(std::true_type, int) { return YdimT; }
                               ^
/home/qon/alice/O2/GPU/GPUTracking/Standalone/TPCFastTransformation/Spline1DSpec.h:286:22: error: use of undeclared identifier 'std'
  static int getYdim(std::false_type, int nYdim) { return nYdim; }
                     ^
/home/qon/alice/O2/GPU/GPUTracking/Standalone/TPCFastTransformation/Spline1DSpec.h:291:11: error: use of undeclared identifier 'std'
  typedef std::integral_constant<bool, YisAbsentT> YisAbsentTbool;
          ^
/home/qon/alice/O2/GPU/GPUTracking/Standalone/TPCFastTransformation/Spline1DSpec.h:292:25: error: use of undeclared identifier 'std'
  static int getMaxYdim(std::true_type, int nYdim) { return nYdim; }
                        ^
/home/qon/alice/O2/GPU/GPUTracking/Standalone/TPCFastTransformation/Spline1DSpec.h:293:35: error: use of undeclared identifier 'std'
  static constexpr int getMaxYdim(std::false_type, int) { return abs(YdimT); }
                                  ^
/home/qon/alice/O2/GPU/GPUTracking/Standalone/TPCFastTransformation/Spline1DSpec.h:308:26: error: use of undeclared identifier 'YisPositiveTbool'; did you mean 'YisPositiveT'?
    auto nYdim = getYdim(YisPositiveTbool{}, inpYdim);
                         ^~~~~~~~~~~~~~~~
                         YisPositiveT                                                                                                                                                                                                                                                                                                                      
/home/qon/alice/O2/GPU/GPUTracking/Standalone/TPCFastTransformation/Spline1DSpec.h:270:43: note: 'YisPositiveT' declared here
template <typename DataT, int YdimT, bool YisPositiveT, bool YisOneT, bool YisAbsentT>
                                          ^
/home/qon/alice/O2/GPU/GPUTracking/Standalone/TPCFastTransformation/Spline1DSpec.h:323:26: error: use of undeclared identifier 'YisPositiveTbool'; did you mean 'YisPositiveT'?
```